acra/security-controls/sql-firewall/pattern_filter.md

```diff
@@ -0,0 +1,139 @@
+---
+title: "Patterns filters"
+bookCollapseSection: true
+---
+
+## Patterns
+
+Patterns work in the following manner: when a query addressing the database is checked, each pattern (if any is detected) is parsed alongside the incoming query. The parsing process extracts the SQL statement type (SELECT, INSERT, UPDATE etc), WHERE and UNION nodes, tables, columns, and values from the body of the query. If the pattern and the query have matching nodes, then tables are checked. If matches for tables are detected, then columns are checked. When separate columns match, a decision to block (through denylist) or to allow (through allowlist) the query is made.
+
+### Supported patterns
+
+{{< hint info >}}This is a work-in-progress area, the list is a subject of fixes and updates.{{< /hint >}}
+
+#### VALUE pattern
+
+`%%VALUE%%` represents literal value (string, binary, number, boolean), it's supported in the following cases: after WHERE, IN, ORDER BY, GROUP BY, BETWEEN.
+
+* **WHERE ... VALUE**
+
+  | pattern | matching queries | non-matching queries |
+  | ------- | ------- | ------- |
+  | `SELECT a `<br/>`FROM t `<br/>`WHERE ID = %%VALUE%%` |  `SELECT a `<br/>`FROM t `<br/>`WHERE ID = '123'`<br/><br/> `SELECT a `<br/>`FROM t `<br/>`WHERE ID = 35`  | `SELECT a, b `<br/>`FROM t `<br/>`WHERE ID = '123'`<br/><br/> `SELECT a `<br/>`FROM anothertable `<br/>`WHERE ID = '123`<br/><br/> `SELECT a FROM t WHERE ID > '123` |
```

Comment:
that is how it will look:
```
<table>
    <thead>
        <tr>
            <td>list of tables</td>
            <td>allowed queries</td>
            <td>denied queries</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td rowspan="2"><code>["EMPLOYEE"]</code></td>
            <td><code>SELECT EMP_ID, LAST_NAME FROM EMPLOYEE WHERE CITY = 'Seattle' ORDER BY EMP_ID;</code></td>
            <td><code>`SELECT EMP_ID, LAST_NAME FROM EMPLOYEE_TBL WHERE CITY = 'Seattle' ORDER BY EMP_ID;</code></td>
        </tr>
        <tr>
            <td><code>INSERT INTO EMPLOYEE values(1, 'email', 'name');</code></td>
            <td><code>INSERT INTO EMPLOYEE_TBL VALUES (1, 'email', 'name');</code></td>
        </tr>
        <tr>
            <td></td>
            <td>
                <code>SELECT EMP_ID, LAST_NAME FROM EMPLOYEE, EMPLOYEE_TBL WHERE CITY = 'Seattle' ORDER BY EMP_ID;</code>
            </td>
        </tr>
    </tbody>
</table>
```
it is example just for one row of examples for `deny` handler. Note, there we should use rowspan for `<td>`. I don't know how will be better. especially if keep in mind that we preferred markdown to allow write docs for people, who don't know html/css...